### ISSUE TITLE
Discuss: set an explicit threshold for test length and reduce the runtime of tests above the threshold

### ISSUE THREAD

**Issue by mr-salty** [googleapis/google-cloud-cpp#3878]
Maybe this should be N separate bugs, but let's just keep playing whack-a-mole with long-running integration tests - these look like the longest ones now:

```
//google/cloud/bigtable/tests:data_integration_test                      PASSED in 104.9s
//google/cloud/storage/tests:bucket_integration_test                     PASSED in 99.6s
//google/cloud/storage/tests:thread_integration_test                     PASSED in 63.9s
//google/cloud/storage/tests:object_insert_integration_test              PASSED in 58.2s
//google/cloud/storage/tests:object_media_integration_test               PASSED in 52.0s
```

**Comment by devjgm** [comment]
What more needs to be done on this issue before we can close it?

**Comment by devjgm** [comment]
@mr-salty Unless there's something you're actively working on related to this issue, I think we should close it. Our builds are looking quite good now. We can file new issues for specific things if slow builds arise.

This is the last issue blocking our closing of the "Optimize CI" milestone.

**Comment by mr-salty** [comment]
I don't think it needs to block the milestone regardless of whether we close the bug or not.

I opened this as a general "when should we consider this 'good enough'?"  I'd like to have some sort of explicit threshold like "no test should take more than 1 minute to run" (just for example - we can debate the exact value).  Ideally we would have something watching the results and filing bugs, but I'm not aware of a tool that does that.

**Comment by coryan** [comment]
> I opened this as a general "when should we consider this 'good enough'?"

I wish you had said that was your intention when you opened the bug, because to me it just looked like "I want to improve these 5 tests".

> I'd like to have some sort of explicit threshold like "no test should take more than 1 minute to run" (just for example - we can debate the exact value). Ideally we would have something watching the results and filing bugs, but I'm not aware of a tool that does that.

Whatever threshold you set, it needs to be different for running against production vs. running against the emulator. Production is several times (maybe as much as 10x) slower.  This is because Production is typically doing more work (like actually persisting data) and because it is "further away".

As to specific thresholds, I also doubt you can set them lower than 300s *per test*.  The spanner integration tests, when running against production, routinely take that long to just create the database where they will run. And they need to create a database or we need to coordinate which test gets to use what database, which is too hard.  Note that often the same integration test may take less than 60s to run, it seems to depend on how much contention there is on a specific spanner instance.

**Comment by mr-salty** [comment]
I meant "whack-a-mole" to mean "keep smacking things down as they pop up", but definitely could have been more explicit - I re-titled the bug.

We do need to be clear about what we're measuring against... I'd say the primary aim is speeding up CI builds, but developer builds are also important (and I could well have that backwards).  5 minutes sounds like a lot in any case.

**Comment by mr-salty** [comment]
FWIW, just one sample (from my workstation, just now) - tests that are above or near 60s and run during `./ci/kokoro/docker/build.sh integration`

```
//google/cloud/spanner/benchmarks:multiple_rows_cpu_benchmark            PASSED in 155.3s
//google/cloud/storage/tests:bucket_integration_test                     PASSED in 132.5s
//google/cloud/bigtable/tests:data_integration_test                      PASSED in 128.0s
//google/cloud/spanner/integration_tests:spanner_client_rpc_failure_threshold_integration_test PASSED in 96.8s
//google/cloud/spanner/samples:samples                                   PASSED in 87.1s
//google/cloud/storage/tests:thread_integration_test                     PASSED in 86.7s
//google/cloud/storage/tests:object_plenty_clients_simultaneously_integration_test PASSED in 70.9s
//google/cloud/storage/tests:object_insert_integration_test              PASSED in 68.8s
//google/cloud/bigtable/tests:mutations_integration_test                 PASSED in 62.1s
//google/cloud/storage/tests:object_media_integration_test               PASSED in 59.6s
//google/cloud/bigtable/tests:filters_integration_test                   PASSED in 55.1s
```

**Comment by devjgm** [comment]
Is there more to be done on this issue, or should it be closed? Or maybe a separate issue spun out w/ a clearer action that needs to be taken?

**Comment by mr-salty** [comment]
We should continue to monitor test execution times although it doesn't seem like there's much enthusiasm for setting a hard threshold.  And, since this hasn't been touched in a long time, I'll just close it.
